x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java

```diff
@@ -136,53 +146,221 @@ protected void readInternal(ByteBuffer b) throws IOException {
+                    final int read;
+                    if ((rangeToRead.v2() - rangeToRead.v1()) < b.remaining()) {
+                        final ByteBuffer duplicate = b.duplicate();
+                        duplicate.limit(duplicate.position() + Math.toIntExact(rangeToRead.v2() - rangeToRead.v1()));
+                        read = readCacheFile(channel, position, duplicate);
+                        assert duplicate.position() <= b.limit();
+                        b.position(duplicate.position());
+                    } else {
+                        read = readCacheFile(channel, position, b);
                     }
+                    return read;
+                }, this::writeCacheFile, directory.cacheFetchAsyncExecutor());
+
+                for (Tuple<Long, Long> indexCacheMiss : indexCacheMisses) {
+                    cacheFile.populateAndRead(indexCacheMiss, indexCacheMiss, channel -> {
+                        final int indexCacheMissLength = Math.toIntExact(indexCacheMiss.v2() - indexCacheMiss.v1());
+
+                        // We assume that we only cache small portions of blobs so that we do not need to:
+                        // - use a BigArrays for allocation
+                        // - use an intermediate copy buffer to read the file in sensibly-sized chunks
+                        // - release the buffer once the indexing operation is complete
+                        assert indexCacheMissLength <= COPY_BUFFER_SIZE : indexCacheMiss;
+
+                        final ByteBuffer byteBuffer = ByteBuffer.allocate(indexCacheMissLength);
+                        Channels.readFromFileChannelWithEofException(channel, indexCacheMiss.v1(), byteBuffer);
+                        // NB use Channels.readFromFileChannelWithEofException not readCacheFile() to avoid counting this in the stats
+                        byteBuffer.flip();
+                        final BytesReference content = BytesReference.fromByteBuffer(byteBuffer);
+                        directory.putCachedBlob(fileInfo.physicalName(), indexCacheMiss.v1(), content);
+                        return indexCacheMissLength;
+                    }, (channel, from, to, progressUpdater) -> {
+                        // normally doesn't happen, we're already obtaining a range covering all cache misses above, but this
+                        // can happen if the real populateAndRead call already failed to obtain this range of the file. In that
+                        // case, we don't retry, we simply fail to populate the index cache.
+                        logger.debug(
+                            "failed to fill index cache miss [{}-{}] of {} due to earlier failure",
+                            from,
+                            to,
+                            CachedBlobContainerIndexInput.this
+                        );
+                        throw new IOException(
```

Comment:
This doesn't happen if we failed to populate the index cache, it only happens if the larger read above already failed.

However it is a good question to ask because I now realise that there may in principle be subsequent reads waiting on this range that we've claimed but then fail to fill. Let's just fill them, see f06e0ae49760a3fd12ca5ff271711df2cb12d0fb.